File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/catalogue/ProductSpecification.java
package com.zappymart.domain.catalogue;

import java.util.List;
import java.util.Locale;

public sealed interface ProductSpecification {

    boolean isSatisfiedBy(Product product);

    record EveryProduct() implements ProductSpecification {
        @Override
        public boolean isSatisfiedBy(Product product) {
            return true;
        }
    }

    record InCategory(String categorySlug) implements ProductSpecification {
        @Override
        public boolean isSatisfiedBy(Product product) {
            return product.category().slug().equals(categorySlug);
        }
    }

    record NameContains(String text) implements ProductSpecification {
        @Override
        public boolean isSatisfiedBy(Product product) {
            return product.name().toLowerCase(Locale.ROOT).contains(text.toLowerCase(Locale.ROOT));
        }
    }

    record InStock() implements ProductSpecification {
        @Override
        public boolean isSatisfiedBy(Product product) {
            return product.isInStock();
        }
    }

    record AllOf(List<ProductSpecification> parts) implements ProductSpecification {
        public AllOf {
            parts = List.copyOf(parts);
        }

        @Override
        public boolean isSatisfiedBy(Product product) {
            return parts.stream().allMatch(part -> part.isSatisfiedBy(product));
        }
    }

    default ProductSpecification and(ProductSpecification other) {
        if (this instanceof EveryProduct) {
            return other;
        }
        if (other instanceof EveryProduct) {
            return this;
        }
        return new AllOf(List.of(this, other));
    }

    static ProductSpecification everyProduct() {
        return new EveryProduct();
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/catalogue/ProductSpecification.kt
package nl.zappymart.domain.catalogue

sealed interface ProductSpecification {

    fun isSatisfiedBy(product: Product): Boolean

    data class InCategory(val categorySlug: String) : ProductSpecification {
        override fun isSatisfiedBy(product: Product) = product.category.slug == categorySlug
    }

    data class NameContains(val text: String) : ProductSpecification {
        override fun isSatisfiedBy(product: Product) = product.name.contains(text, ignoreCase = true)
    }

    data object InStock : ProductSpecification {
        override fun isSatisfiedBy(product: Product) = product.stock > 0
    }

    data class MatchingAll(val parts: List<ProductSpecification>) : ProductSpecification {
        override fun isSatisfiedBy(product: Product) = parts.all { part -> part.isSatisfiedBy(product) }
    }

    companion object {
        val EVERYTHING = MatchingAll(emptyList())

        fun of(categorySlug: String?, nameContains: String?, inStockOnly: Boolean?): ProductSpecification {
            val parts = buildList {
                categorySlug?.let { slug -> add(InCategory(slug)) }
                nameContains?.takeIf { text -> text.isNotBlank() }?.let { text -> add(NameContains(text.trim())) }
                if (inStockOnly == true) add(InStock)
            }
            return MatchingAll(parts)
        }
    }
}
